### PR TITLE
Hardware Management, Metrics, PWM/GPIO support.

### PR DESCRIPTION
**Pull request by XzibitGG**

### CODE REVIEW

**Comment by Bankst** [comment]
Waiting on @mcm001 and myself to finish the bootup-sprint.

**Comment by Bankst** [comment]
[Hardware interaction system](https://trello.com/c/H6YlnY2j/12-hardware-interaction-system)

File: photon-server/src/main/java/org/photonvision/common/hardware/PWM/CustomPWM.java [hunk 1]
```diff
@@ -18,10 +18,11 @@
 package org.photonvision.common.hardware.PWM;
 
 import org.photonvision.common.configuration.HardwareConfig;
+import org.photonvision.common.hardware.Platform;
 
 public class CustomPWM extends PWMBase {
     private int pwmRate = 0;
-    private int[] pwmRange = new int[2];
+    private int pwmRange = 0;
```
Comment (Bankst) [inline photon-server/src/main/java/org/photonvision/common/hardware/PWM/CustomPWM.java:25]: Why did this change from an array? It's possible to want a lower limit to the range

File: photon-server/src/main/java/org/photonvision/common/configuration/HardwareConfig.java [hunk 2]
```diff
@@ -17,7 +17,142 @@
...
+    private final String ledDimCommand;
+    private final String ledBlinkCommand;
+
+    // Metrics
+    private final String cpuTempCommand;
+    private final String cpuMemoryCommand;
+    private final String cpuUtilCommand;
+    private final String gpuMemoryCommand;
+    private final String gpuTempCommand;
+    private final String ramUtilCommand;
```
Comment (mcm001) [inline photon-server/src/main/java/org/photonvision/common/configuration/HardwareConfig.java:47]: These are all `final`, so maybe we can replace with getters?

File: photon-server/src/main/java/org/photonvision/common/hardware/GPIO/GPIOBase.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                }
+            };
+
+    private static final ShellExec runCommand = new ShellExec(true, true);
+
+    public static String execute(String command) {
+        try {
+            runCommand.executeBashCommand(command);
+        } catch (Exception e) {
+            return "";
```
Comment (Bankst) [inline photon-server/src/main/java/org/photonvision/common/hardware/GPIO/GPIOBase.java:39]: Log exceptions here

File: photon-server/src/main/java/org/photonvision/common/hardware/Platform.java [hunk 4]
```diff
@@ -15,25 +15,28 @@
...
-    UNSUPPORTED(
-            "Unsupported Platform - OS: "
-                    + SystemUtils.OS_NAME
-                    + ", Architecture: "
-                    + SystemUtils.OS_ARCH);
+    LINUX_RASPBIAN("Linux Raspbian"), // Raspberry Pi 3/4
+    LINUX_AARCH64BIONIC("Linux AARCH64 Bionic"), // Jetson Nano, Jetson TX2
+    MACOS_64("Mac OS x64"),
+
+    // ChameleonVision Supported (Manual install)
```
Comment (mcm001) [inline photon-server/src/main/java/org/photonvision/common/hardware/Platform.java:34]: ChameleonVision :bolb:

File: photon-server/src/main/java/org/photonvision/common/hardware/Platform.java [hunk 5]
```diff
@@ -42,14 +45,23 @@
...
+    }
+
+    public boolean isLinux() {
+        return this == LINUX_64 || this == LINUX_RASPBIAN || this == LINUX_ARM64;
     }
 
-    public static boolean isLinux() {
-        return CurrentPlatform != UNSUPPORTED && !isWindows();
+    public boolean isMac() {
+        return this == MACOS_64;
```
Comment (mcm001) [inline photon-server/src/main/java/org/photonvision/common/hardware/Platform.java:64]: CSCore doesn't support mac, so I don't think we need this.

File: photon-server/src/main/java/org/photonvision/common/hardware/Platform.java [hunk 6]
```diff
@@ -87,17 +99,26 @@ private static Platform getCurrentPlatform() {
             if (RuntimeDetector.is64BitIntel()) return WINDOWS_64;
         }
 
+        if (RuntimeDetector.isMac()) {
+            if (RuntimeDetector.is32BitIntel()) return UNSUPPORTED;
+            if (RuntimeDetector.is64BitIntel()) return MACOS_64;
+        }
```
Comment (mcm001) [inline photon-server/src/main/java/org/photonvision/common/hardware/Platform.java:105]: Same as above

File: photon-server/src/main/java/org/photonvision/common/hardware/metrics/MetricsPublisher.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        var gpu = GPU.getInstance();
+        var ram = RAM.getInstance();
+
+        metrics = new HashMap<>();
+
+        this.metricsThread =
+                new Thread(
+                        () -> {
+                            var timer = new Timer();
+                            timer.schedule(
```
Comment (mcm001) [inline photon-server/src/main/java/org/photonvision/common/hardware/metrics/MetricsPublisher.java:46]: Just a note, let's maybe make a central thread pool and add executors for it.

File: photon-server/src/test/resources/hardware/HardwareConfig.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  "metrics": {
+    "cpuTemp": "echo 10",
+    "cpuMemory": "echo 10",
+    "cpuUtil": "echo 10",
+    "gpuMemory": "echo 10",
+    "gpuUtil": "echo 10",
+    "ramUtil": "echo 10"
+  }
+}
```
Comment (mcm001) [inline photon-server/src/test/resources/hardware/HardwareConfig.json:23]: EOF return?

File: photon-server/src/main/java/org/photonvision/common/hardware/HardwareManager.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import org.photonvision.common.hardware.PWM.PiPWM;
+import org.photonvision.common.hardware.metrics.MetricsBase;
+import org.photonvision.common.hardware.metrics.MetricsPublisher;
+
+public class HardwareManager {
+    HardwareConfig hardwareConfig;
+    private static final HashMap<Integer, Pair<? extends GPIOBase, ? extends PWMBase>> LEDs =
+            new HashMap<>();
+
+    public static HardwareManager getInstance() {
```
Comment (Bankst) [inline photon-server/src/main/java/org/photonvision/common/hardware/HardwareManager.java:37]: Use thread-safe singleton pattern.
```java
private static class Singleton {
    public static final HardwareManager INSTANCE = new HardwareManager();
}

public static HardwareManager getInstance() {
    return Singleton.INSTANCE;
}

File: photon-server/src/main/java/org/photonvision/common/hardware/PWM/PWMBase.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                }
+            };
+
+    private static final ShellExec runCommand = new ShellExec(true, true);
+
+    public static String execute(String command) {
+        try {
+            runCommand.executeBashCommand(command);
+        } catch (Exception e) {
+            return "";
```
Comment (Bankst) [inline photon-server/src/main/java/org/photonvision/common/hardware/PWM/PWMBase.java:40]: Log exceptions

File: photon-server/src/main/java/org/photonvision/common/util/TestUtils.java [hunk 11]
```diff
@@ -165,6 +166,10 @@ public static Path getDotBoardImagesPath() {
         return getResourcesFolderPath().resolve("calibrationBoardImages");
     }
 
+    public static File getHardwareConfigJson() {
+        return getResourcesFolderPath().resolve("hardware").resolve("HardwareConfig.json").toFile();
```
Comment (Bankst) [inline photon-server/src/main/java/org/photonvision/common/util/TestUtils.java:170]: move to root folder

Comment (XzibitGG) [reply photon-server/src/main/java/org/photonvision/common/util/TestUtils.java:170]: This is getting the test config json file for unit tests. Not sure how I'd move it to the root folder.

**Pull request merged**